### PR TITLE
HPCC-24027 Ensure thormaster exception is cleared by agent.

### PR DESCRIPTION
**Pull request by jakesmith**
Clear the exception passed back from a k8s wuid job, so that
if the workunit is submitted (e.g. due to RECOVERY) it will not
cause the same exception to be re-seen.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by jakesmith** [comment]
@ghalliday - please review

Re. whitespaces (trimmed trailing), this is being caused by an option I have set in VSCode, to prevent trailing white spaces, which automatically strips any trailing whitespaces from end of lines when a file is saved.
It's convenient and avoids them being introduced and clears up the existing erroneous spaces on a per-file basis. Let me know if you okay with that in this/other commits, e.g. by reviewing in github by turning off white spaces from the diff.

File: ecl/eclagent/eclgraph.cpp [hunk 1]
```diff
@@ -1590,7 +1591,12 @@ void EclAgent::executeThorGraph(const char * graphName)
             {
                 str.clear();
                 e.getExceptionMessage(str);
-                throw makeStringException(e.getExceptionCode(), str.str());
+                unsigned errCode = e.getExceptionCode();
+                {
+                    WorkunitUpdate w = updateWorkUnit();
```
Comment (ghalliday) [inline ecl/eclagent/eclgraph.cpp:1596]: could do with a comment explaining why this is being removed. (and why the reload is required above)

**Comment by ghalliday** [comment]
@jakesmith code looks fine (a shame it is cloned).  Could do with a comment though to explain why it is reloading and removing.

**Comment by richardkchapman** [comment]
Not clear why this would be k8s-specific?

**Comment by jakesmith** [comment]
>Not clear why this would be k8s-specific?

In bare-metal setup, the agent connects to Thor via a socket, and receives this exception back as a reply. In container/k8s mode, there isn't a similar connection, so the conduit for the return error is a workunit exception.

**Comment by jakesmith** [comment]
> (a shame it is cloned).

agree, have thought about combining before, + it's caused problems before because a change was forgotten in one vs the other.

Decided to combine now, they were identical really.

> Could do with a comment though to explain why it is reloading and removing.

Have added comments.

@ghalliday - please review

**Comment by ghalliday** [comment]
@jakesmith looks good, please squash.

**Comment by jakesmith** [comment]
@ghalliday - squashed

**Comment by jakesmith** [comment]
@ghalliday - squashed and rebased

**Pull request merged**